### PR TITLE
Update ZeroExApiAdapter to support multiplex

### PR DESCRIPTION
**Pull request by cece-z**
# Code Review Processes
## New Feature Review
Before submitting a pull request for new review, make sure the following is done:
* Design doc is created and posted here: [Insert Link]
* Code cleanliness and completeness is addressed via [guidelines](https://app.gitbook.com/@setprotocol-1/s/set/smart-contract-engineering/sc-code-review-process)

README Checks
- [x] README has proper context for the reviewer to understand what the code includes, any important design considerations, and areas to pay more attention to

Code Checks
- [x] Add explanatory comments. If there is complex code that requires specific context or understanding, note that in a comment
- [x] Remove unncessary comments. Any comments that do not add additional context, information, etc. should be removed
- [] Add javadocs. 
- [x] Scrub through the code for inconsistencies (e.g. removing extra spaces)
- [x] Ensure there are not any .onlys in spec files

Broader Considerations
- [x] Ensure variable, function and event naming is clear, consistent, and reflective for the scope of the code.
- [x] Consider if certain pieces of logic should be placed in a different library, module

### CODE REVIEW

File: contracts/protocol/integration/exchange/ZeroExApiAdapter.sol [hunk 1]
```diff
@@ -201,7 +214,58 @@ contract ZeroExApiAdapter {
                 (inputToken, outputToken) = _decodeTokensFromUniswapV3EncodedPath(encodedPath);
                 require(inputToken == wethAddress, "First token must be WETH");
                 inputToken = ETH_ADDRESS;
-            } else {
+            } else if (selector == 0xf35b4733) {
+                // multiplexBatchSellEthForToken()
+                MultiplexBatchFillData memory fillData;
+                (fillData, minOutputTokenAmount) =
+                	abi.decode(_data[4:], (MultiplexBatchFillData, uint256));
```
Comment (dorothy-zbornak) [inline contracts/protocol/integration/exchange/ZeroExApiAdapter.sol:221]: The function is [defined as](https://github.com/0xProject/protocol/blob/development/contracts/zero-ex/contracts/src/features/interfaces/IMultiplexFeature.sol#L124)
```
    function multiplexBatchSellEthForToken(
        IERC20TokenV06 outputToken,
        BatchSellSubcall[] calldata calls,
        uint256 minBuyAmount
    )
```
so we really want something like
```suggestion
                BatchSellSubcall[] memory subcalls;
                (outputToken, subcalls, minOutputTokenAmount) =
                	abi.decode(_data[4:], (address, BatchSellSubCall[], uint256));
```

But here's a thing:  we don't actually care about what's inside the `subcalls` array so we can trick the decoder into thinking it's just a plain uint256 (dynamic types are serialized as uint256 offsets) and save a ton of gas. So we can just do:
```suggestion
                (outputToken, , minOutputTokenAmount) =
                	abi.decode(_data[4:], (address, uint256, uint256));
```

File: contracts/protocol/integration/exchange/ZeroExApiAdapter.sol [hunk 2]
```diff
@@ -201,7 +214,58 @@ contract ZeroExApiAdapter {
...
+                inputToken = ETH_ADDRESS;
+                outputToken = fillData.outputToken;
+                inputTokenAmount = _sourceQuantity;
+			} else if (selector == 0x77725df6) {
+                // multiplexBatchSellTokenForEth()
+                MultiplexBatchFillData memory fillData;
+                (fillData, minOutputTokenAmount) =
+                	abi.decode(_data[4:], (MultiplexBatchFillData, uint256));
```
Comment (dorothy-zbornak) [inline contracts/protocol/integration/exchange/ZeroExApiAdapter.sol:229]: And [this one](https://github.com/0xProject/protocol/blob/development/contracts/zero-ex/contracts/src/features/interfaces/IMultiplexFeature.sol#L141) is
```
    function multiplexBatchSellTokenForEth(
        IERC20TokenV06 inputToken,
        BatchSellSubcall[] calldata calls,
        uint256 sellAmount,
        uint256 minBuyAmount
    )
```
so (using the same trick)
```suggestion
                (inputToken, , inputTokenAmount, minOutputTokenAmount) =
                	abi.decode(_data[4:], (address, uint256, uint256, uint256));
```

File: contracts/protocol/integration/exchange/ZeroExApiAdapter.sol [hunk 3]
```diff
@@ -201,7 +214,58 @@ contract ZeroExApiAdapter {
...
+                inputToken = fillData.inputToken;
+                outputToken = ETH_ADDRESS;
+                inputTokenAmount = fillData.sellAmount;
+			} else if (selector == 0x7a1eb1b9) {
+                // multiplexBatchSellTokenForToken()
+                MultiplexBatchFillData memory fillData;
+                (fillData, minOutputTokenAmount) =
+                	abi.decode(_data[4:], (MultiplexBatchFillData, uint256));
```
Comment (dorothy-zbornak) [inline contracts/protocol/integration/exchange/ZeroExApiAdapter.sol:237]: This one is
```
    function multiplexBatchSellTokenForToken(
        IERC20TokenV06 inputToken,
        IERC20TokenV06 outputToken,
        BatchSellSubcall[] calldata calls,
        uint256 sellAmount,
        uint256 minBuyAmount
    )
```
so
```suggestion
                (inputToken, outputToken, , inputTokenAmount, minOutputTokenAmount) =
                	abi.decode(_data[4:], (address, address, uint256, uint256, uint256));
```

File: contracts/protocol/integration/exchange/ZeroExApiAdapter.sol [hunk 4]
```diff
@@ -201,7 +214,58 @@ contract ZeroExApiAdapter {
...
+                inputToken = fillData.inputToken;
+                outputToken = fillData.outputToken;
+                inputTokenAmount = fillData.sellAmount;
+			} else if (selector == 0x5161b966) {
+                // multiplexMultiHopSellEthForToken()
+                MultiHopFillData memory fillData;
+                (fillData, minOutputTokenAmount) =
+                	abi.decode(_data[4:], (MultiHopFillData, uint256));
```
Comment (dorothy-zbornak) [inline contracts/protocol/integration/exchange/ZeroExApiAdapter.sol:245]: And this one is
```
    function multiplexMultiHopSellEthForToken(
        address[] calldata tokens,
        MultiHopSellSubcall[] calldata calls,
        uint256 minBuyAmount
    )
```
and here we also don't care about what's inside the `calls` so we can effectively skip decoding it by pretending it's a plain `uint256`:
```suggestion
                address[] memory tokens;
                (tokens, , minOutputTokenAmount) =
                	abi.decode(_data[4:], (address[], uint256, uint256));
```
You can derive `inputToken` and `outputToken` from the `tokens` array (first and last entries, respectively).

File: contracts/protocol/integration/exchange/ZeroExApiAdapter.sol [hunk 5]
```diff
@@ -201,7 +214,58 @@ contract ZeroExApiAdapter {
...
+                inputTokenAmount = fillData.sellAmount;
+			} else if (selector == 0x0f3b31b2) {
+                // multiplexMultiHopSellTokenForToken()
+                MultiHopFillData memory fillData;
+                (fillData, minOutputTokenAmount) =
+                	abi.decode(_data[4:], (MultiHopFillData, uint256));
+                require(fillData.tokens.length > 1, "Multihop token path too short");
+                inputToken = fillData.tokens[0];
+                outputToken = fillData.tokens[fillData.tokens.length - 1];
+                inputTokenAmount = fillData.sellAmount;
```
Comment (dorothy-zbornak) [inline contracts/protocol/integration/exchange/ZeroExApiAdapter.sol:267]: I think you get the idea for the rest of these.

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 6]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+
+    describe("multiplexMultiHopSellEthForToken", () => {
+      it("validates data", async () => {
+        const data = zeroExMock.interface.encodeFunctionData("multiplexMultiHopSellEthForToken", [
+          [sourceToken, destToken],
+          ZERO,
+          minDestinationQuantity,
+        ]);
+        const [target, value, _data] = await zeroExApiAdapter.getTradeCalldata(
+          sourceToken,
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1057]:
```suggestion
          ethToken,
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 7]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+      });
+
+      it("rejects wrong input token", async () => {
+        const data = zeroExMock.interface.encodeFunctionData("multiplexMultiHopSellEthForToken", [
+          [sourceToken, destToken],
+          ZERO,
+          minDestinationQuantity,
+        ]);
+        const tx = zeroExApiAdapter.getTradeCalldata(
+          sourceToken,
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1076]:
```suggestion
          ethToken,
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 8]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          destToken,
+          destination,
+          sourceQuantity,
+          minDestinationQuantity,
+          data,
+        );
+        await expect(tx).to.be.revertedWith("Mismatched input token");
+      });
+
+      it("rejects went path too short", async () => {
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1086]:
```suggestion
      it("rejects when path too short", async () => {
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 9]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+        const data = zeroExMock.interface.encodeFunctionData("multiplexMultiHopSellEthForToken", [
+          [sourceToken, destToken],
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1088]:
```suggestion
          [ethToken],
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 10]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          ZERO,
+          minDestinationQuantity,
+        ]);
+        const tx = zeroExApiAdapter.getTradeCalldata(
+          sourceToken,
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1093]:
```suggestion
          ethToken,
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 11]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          sourceQuantity,
+          minDestinationQuantity,
+          data,
+        );
+        await expect(tx).to.be.revertedWith("Multihop token path too short");
+      });
+
+      it("rejects wrong output token", async () => {
+        const data = zeroExMock.interface.encodeFunctionData("multiplexMultiHopSellEthForToken", [
+          [sourceToken, destToken],
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1105]:
```suggestion
          [ethToken, destToken],
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 12]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          ZERO,
+          minDestinationQuantity,
+        ]);
+        const tx = zeroExApiAdapter.getTradeCalldata(
+          sourceToken,
+          destToken,
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1111]:
```suggestion
          ethToken,
          otherToken,
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 13]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          destination,
+          sourceQuantity,
+          minDestinationQuantity,
+          data,
+        );
+        await expect(tx).to.be.revertedWith("Mismatched output token");
+      });
+
+      it("rejects wrong input token quantity", async () => {
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1120]: Not testable

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 14]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          sourceQuantity,
+          minDestinationQuantity,
+          data,
+        );
+        await expect(tx).to.be.revertedWith("Mismatched input token quantity");
+      });
+
+      it("rejects wrong output token quantity", async () => {
+        const data = zeroExMock.interface.encodeFunctionData("multiplexMultiHopSellEthForToken", [
+          [sourceToken, destToken],
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1139]:
```suggestion
          [ethToken, destToken],
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 15]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          ZERO,
+          minDestinationQuantity,
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1141]:
```suggestion
          otherQuantity,
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 16]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+        ]);
+        const tx = zeroExApiAdapter.getTradeCalldata(
+          sourceToken,
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1144]:
```suggestion
          ethToken,
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 17]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          data,
+        );
+        await expect(tx).to.be.revertedWith("Mismatched output token quantity");
+      });
+    });
+
+    describe("multiplexMultiHopSellTokenForEth", () => {
+      it("validates data", async () => {
+        const data = zeroExMock.interface.encodeFunctionData("multiplexMultiHopSellTokenForEth", [
+          [sourceToken, destToken],
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1158]:
```suggestion
          [sourceToken, ethToken],
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 18]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          ZERO,
+          sourceQuantity,
+          minDestinationQuantity,
+        ]);
+        const [target, value, _data] = await zeroExApiAdapter.getTradeCalldata(
+          sourceToken,
+          destToken,
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1165]:
```suggestion
          ethToken,
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 19]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          data,
+        );
+        expect(target).to.eq(zeroExMock.address);
+        expect(value).to.deep.eq(ZERO);
+        expect(_data).to.deep.eq(data);
+      });
+
+      it("rejects wrong input token", async () => {
+        const data = zeroExMock.interface.encodeFunctionData("multiplexMultiHopSellTokenForEth", [
+          [sourceToken, destToken],
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1178]:
```suggestion
          [otherToken, ethToken],
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 20]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          ZERO,
+          sourceQuantity,
+          minDestinationQuantity,
+        ]);
+        const tx = zeroExApiAdapter.getTradeCalldata(
+          sourceToken,
+          destToken,
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1185]:
```suggestion
          sourceToken,
          ethToken,
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 21]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          sourceQuantity,
+          minDestinationQuantity,
+          data,
+        );
+        await expect(tx).to.be.revertedWith("Mismatched input token");
+      });
+
+      it("rejects went path too short", async () => {
+        const data = zeroExMock.interface.encodeFunctionData("multiplexMultiHopSellTokenForEth", [
+          [sourceToken, destToken],
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1196]:
```suggestion
          [ethToken],
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 22]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          ZERO,
+          sourceQuantity,
+          minDestinationQuantity,
+        ]);
+        const tx = zeroExApiAdapter.getTradeCalldata(
+          sourceToken,
+          destToken,
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1203]:
```suggestion
          ethToken,
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 23]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+      it("rejects wrong output token", async () => {
+        const data = zeroExMock.interface.encodeFunctionData("multiplexMultiHopSellTokenForEth", [
+          [sourceToken, destToken],
+          ZERO,
+          sourceQuantity,
+          minDestinationQuantity,
+        ]);
+        const tx = zeroExApiAdapter.getTradeCalldata(
+          sourceToken,
+          destToken,
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1221]:
```suggestion
          ethToken,
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 24]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          data,
+        );
+        await expect(tx).to.be.revertedWith("Mismatched output token");
+      });
+
+      it("rejects wrong input token quantity", async () => {
+        const data = zeroExMock.interface.encodeFunctionData("multiplexMultiHopSellTokenForEth", [
+          [sourceToken, destToken],
+          ZERO,
+          sourceQuantity,
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1234]:
```suggestion
          [sourceToken, ethToken],
          ZERO,
          otherQuantity,
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 25]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          minDestinationQuantity,
+        ]);
+        const tx = zeroExApiAdapter.getTradeCalldata(
+          sourceToken,
+          destToken,
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1239]:
```suggestion
          ethToken,
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 26]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          sourceQuantity,
+          minDestinationQuantity,
+          data,
+        );
+        await expect(tx).to.be.revertedWith("Mismatched input token quantity");
+      });
+
+      it("rejects wrong output token quantity", async () => {
+        const data = zeroExMock.interface.encodeFunctionData("multiplexMultiHopSellTokenForEth", [
+          [sourceToken, destToken],
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1250]:
```suggestion
          [sourceToken, ethToken],
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 27]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          ZERO,
+          sourceQuantity,
+          otherQuantity,
+        ]);
+        const tx = zeroExApiAdapter.getTradeCalldata(
+          sourceToken,
+          destToken,
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1257]:
```suggestion
          ethToken,
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 28]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          data,
+        );
+        expect(target).to.eq(zeroExMock.address);
+        expect(value).to.deep.eq(ZERO);
+        expect(_data).to.deep.eq(data);
+      });
+
+      it("rejects wrong input token", async () => {
+        const data = zeroExMock.interface.encodeFunctionData("multiplexMultiHopSellTokenForToken", [
+          [sourceToken, destToken],
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1290]:
```suggestion
          [otherToken, destToken],
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 29]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          sourceQuantity,
+          minDestinationQuantity,
+          data,
+        );
+        await expect(tx).to.be.revertedWith("Mismatched input token");
+      });
+
+      it("rejects went path too short", async () => {
+        const data = zeroExMock.interface.encodeFunctionData("multiplexMultiHopSellTokenForToken", [
+          [sourceToken, destToken],
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1308]:
```suggestion
          [sourceToken],
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 30]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          data,
+        );
+        await expect(tx).to.be.revertedWith("Mismatched output token");
+      });
+
+      it("rejects wrong input token quantity", async () => {
+        const data = zeroExMock.interface.encodeFunctionData("multiplexMultiHopSellTokenForToken", [
+          [sourceToken, destToken],
+          ZERO,
+          sourceQuantity,
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1346]:
```suggestion
          otherQuantity,
```

File: test/protocol/integration/exchange/zeroExApiAdapter.spec.ts [hunk 31]
```diff
@@ -664,7 +763,620 @@ describe("ZeroExApiAdapter", () => {
...
+          minDestinationQuantity,
+          data,
+        );
+        await expect(tx).to.be.revertedWith("Multihop token path too short");
+      });
+
+      it("rejects wrong output token", async () => {
+        const data = zeroExMock.interface.encodeFunctionData("multiplexMultiHopSellTokenForToken", [
+          [sourceToken, destToken],
```
Comment (dorothy-zbornak) [inline test/protocol/integration/exchange/zeroExApiAdapter.spec.ts:1326]:
```suggestion
          [sourceToken, otherToken],
```

**Review by mzhu25 — APPROVED** [review]
lgtm 👍

**Comment by bweick** [comment]
Let's remove the ETH related functions...we don't want anyone dealing directly with ETH since our system doesn't support it.

**Comment by bweick** [comment]
Also if you wouldn't mind bumping the package version to `0.1.1` that would be great so we can export these changes.

**Comment by cece-z** [comment]
@bweick Thanks. Should I also remove two old functions `sellTokenForEthToUniswapV3()` and `sellEthForTokenToUniswapV3()` as well?

**Comment by bweick** [comment]
> @bweick Thanks. Should I also remove two old functions `sellTokenForEthToUniswapV3()` and `sellEthForTokenToUniswapV3()` as well?

yes let's

**Review by bweick — APPROVED** [review]

**Pull request merged**